pmd-core/src/main/java/net/sourceforge/pmd/cpd/Tokens.java

```diff
@@ -5,42 +5,154 @@
 package net.sourceforge.pmd.cpd;
 
 import java.util.ArrayList;
-import java.util.Iterator;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
 
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.annotation.InternalApi;
+import net.sourceforge.pmd.lang.ast.TokenMgrError;
+import net.sourceforge.pmd.lang.document.FileId;
+import net.sourceforge.pmd.lang.document.TextDocument;
+
+/**
+ * Global token collector for CPD. This is populated by lexing all files,
+ * after which the match algorithm proceeds.
+ */
+@InternalApi
```

Comment:
we should probably just move this to a `cpd.internal` package.

Reply:
I think it's fine to leave it here, as it allows us to have some package protected methods. I think the cpd package is more cohesive that way too.

Reply:
My concern with that is that it will still be included in javadoc, and it is not marked as deprecated, so anyone may stumble onto this and use it if they are not extra careful.

So far the use of this annotation has been accompanied by @Deprecated / @DeprecatedUntil700 to issue a very visible warning and let us now we don't want to remove the element, simply internalize it. So third party developers are being signaled that APIs are safe to use as long as they public, are not deprecated, and are not part of an internal package.

With this we would be including a fourth rule, and one that has no direct hint from the IDE / would not be evident in a PR, but needs manual checking. Therefore, it seems error prone.

Reply:
> So far the use of this annotation has been accompanied by @deprecated / @DeprecatedUntil700 to issue a very visible warning and let us now we don't want to remove the element, simply internalize it.

There are actually many existing APIs that only have `@InternalApi` and not `@Deprecated`. Same for `@Experimental`. Using packaging as the sole way of marking internal API is pretty inflexible IMHO.

> So third party developers are being signaled that APIs are safe to use as long as they public, are not deprecated, and are not part of an internal package.

I would add "and so long as they are not `@InternalApi` or `@Experimental`" to this. I don't see why these annotations must be accompanied by `@Deprecated` if we're not going to remove them. In fact it would probably be confusing to us that these be always together, because we wouldn't know if they need to be removed or not. When we upgrade from java 8 we can use `@Deprecated(forRemoval = false)`, although an API that is both Experimental and Deprecated seems contradictory to me, semantically.


That being said, in this specific case `Tokens` can probably be made entirely package private... I'll look into it

Reply:
This should be revisited with #4348